metagov/metagov/core/identity.py

```diff
@@ -0,0 +1,239 @@
+from contextlib import suppress
+from enum import Enum
+
+import json, random
+from django.db import models, IntegrityError
+
+from metagov.core.models import Community
+
+
+class MetagovID(models.Model):
+    """Metagov ID table links all public_ids to a single internal representation of a user. When data
+    associated with public_ids conflicts, primary_ID is used.
+
+    Fields:
+
+    community: foreign key - metagov community the user is part of
+    internal_id: integer - unique, secret ID
+    public_ids: list of integers - public IDs shared with Driver, callers, etc. must have at least one item
+    primary_id: integer - used to resolve conflicts if there are multiple public_ids associated with a user
+
+    FIXME: I don't like that linked_accounts links to this model as metagov_id but it's actually
+    linking to one of the public_ids (which, admittedly should resolve to a single metagov_id)"""
+
+    community = models.ForeignKey(Community)
+    internal_id = models.PositiveIntegerField(unique=True)
+    public_ids = models.JSONField(default=list)
+    primary_id = models.PositiveIntegerField(unique=True)
+
+    # Internal logic
+
+    def add_id(self, new_id):
+        if not self.public_ids:
+            self.primary_id = new_id
+        self.public_ids.append(new_id)
+
+    def change_primary_id(self, new_primary_id):
+        if new_primary_id not in self.public_ids:
+            raise ValueError(f"New primary_id {new_primary_id} not in public IDs")
+        self.primary_id = new_primary_id
+
+
+class LinkType(Enum):
+    OAUTH = "oauth"
+    MANUAL_ADMIN = "manual admin"
+    EMAIL_MATCHING = "email matching"
+    UNKNOWN = "unknown"
+
+
+class LinkQuality(Enum):
+    STRONG_CONFIRM = "confirmed (strong)"
+    WEAK_CONFIRM = "confirmed (weak)"
+    UNCONFIRMED = "unconfirmed"
+    UNKNOWN = "unknown"
+
+
+class LinkedAccounts(models.Model):
```

Comment:
small thing: renaming to `LinkedAccount` since each record only represents one account